src/rpc/net.cpp

```diff
@@ -320,20 +322,31 @@ static RPCHelpMan addconnection()
         throw std::runtime_error("addconnection is for regression testing (-regtest mode) only");
     }
 
-    RPCTypeCheckArgument(request.params, UniValue::VSTR);
+    RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VSTR});
     const std::string address = request.params[0].get_str();
+    const std::string conn_type_in = request.params[1].get_str();
+    ConnectionType conn_type{};
+    if (conn_type_in == "outbound") {
+        conn_type = ConnectionType::OUTBOUND_FULL_RELAY;
+    } else if (conn_type_in == "blockrelay") {
+        conn_type = ConnectionType::BLOCK_RELAY;
+    } else {
+        throw std::runtime_error(self.ToString());
+    }
+
     NodeContext& context = EnsureNodeContext(request.context);
     if (!context.connman) {
         throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
     }
 
-    const bool success = context.connman->AddConnection(address);
+    const bool success = context.connman->AddConnection(address, conn_type);
     if (!success) {
-        throw JSONRPCError(RPC_CLIENT_NODE_CAPACITY_REACHED, "Error: Max number of outbound full relay connections already open.");
+        throw JSONRPCError(RPC_CLIENT_NODE_CAPACITY_REACHED, "Error: Already at capacity for specified connection type.");
```

Comment:
0fb4635bbabb1bea10a215bde4c803ceb5bc5f26
Does this error message describe not having `grant` in the `AddConnection` function?

Reply:
depends for what type. `AddConnection` goes through the nodes and counts how many full relays & block relays there currently are, then compares this to their respective max. it also checks the grant for the total outbound grants. if any of these fail, this error message is hit

Reply:
Same as https://github.com/bitcoin/bitcoin/pull/19315/files#r533943505, consider making this more detailed?  I could see this being misleading if the RPC fails because of max total outbound capacity, as opposed to max capacity for the given type.

Reply:
it's hard to hit the total max without hitting the max for the type :) 

the max values are 
outbound: 8
block-relay-only: 2
total: 10

Reply:
This seems like an appropriate amount of detail for an error message (especially one for an RPC which is test-only).